bitcoin/script/miniscript.h

```diff
@@ -1014,224 +1014,384 @@ bool DecomposeScript(const CScript& script, std::vector<std::pair<opcodetype, st
+            constructed.pop_back();
+            NodeRef<Key> right = std::move(constructed.back());
+            constructed.pop_back();
+            constructed.push_back(MakeNodeRef<Key>(NodeType::OR_C, Vector(std::move(left), std::move(right))));
+            break;
+        }
+        case DecodeContext::OrD: {
+            NodeRef<Key> left = std::move(constructed.back());
+            constructed.pop_back();
+            NodeRef<Key> right = std::move(constructed.back());
+            constructed.pop_back();
+            constructed.push_back(MakeNodeRef<Key>(NodeType::OR_D, Vector(std::move(left), std::move(right))));
+            break;
+        }
+        case DecodeContext::AndOr: {
+            NodeRef<Key> left = std::move(constructed.back());
+            constructed.pop_back();
+            NodeRef<Key> right = std::move(constructed.back());
+            constructed.pop_back();
+            NodeRef<Key> mid = std::move(constructed.back());
+            constructed.pop_back();
+            constructed.push_back(MakeNodeRef<Key>(NodeType::ANDOR, Vector(std::move(left), std::move(mid), std::move(right))));
+            break;
+        }
+        case DecodeContext::ThreshW: {
+            if (in >= last) return {};
+            if (in[0].first == OP_ADD) {
+                ++in;
+                to_parse.emplace_back(DecodeContext::ThreshW, n+1, k);
+                to_parse.emplace_back(DecodeContext::WExpression, -1, -1);
+            } else {
+                to_parse.emplace_back(DecodeContext::ThreshE, n+1, k);
+                to_parse.emplace_back(DecodeContext::Expression, -1, -1);
+            }
+            break;
+        }
+        case DecodeContext::ThreshE: {
+            std::vector<NodeRef<Key>> subs;
+            for (int i = 0; i < n; ++i) {
+                NodeRef<Key> sub = std::move(constructed.back());
+                constructed.pop_back();
+                subs.push_back(std::move(sub));
+            }
+            constructed.push_back(MakeNodeRef<Key>(NodeType::THRESH, std::move(subs), k));
+            break;
+        }
+        case DecodeContext::EndIf: {
+            if (in >= last) return {};
 
-//! Decode a list of script elements into a miniscript (only a: and s:)
-template<typename Key, typename Ctx, typename I>
-inline NodeRef<Key> DecodeWrapped(I& in, I last, const Ctx& ctx) {
-    if (last - in >= 3 && in[0].first == OP_FROMALTSTACK) {
-        ++in;
-        auto sub = DecodeMulti<Key>(in, last, ctx);
-        if (!sub) return {};
-        if (in == last || in[0].first != OP_TOALTSTACK) return {};
-        ++in;
-        return MakeNodeRef<Key>(NodeType::WRAP_A, Vector(std::move(sub)));
+            // could be andor or or_i
+            if (in[0].first == OP_ELSE) {
+                ++in;
+                to_parse.emplace_back(DecodeContext::EndIfElse, -1, -1);
+                to_parse.emplace_back(DecodeContext::MaybeAndV, -1, -1);
```

Comment:
The MaybeAndV + Expression combination occurs a few times. Perhaps it makes sense to add a meta-context (like the `DecodeMulti` there was before) that is just those two? Feel free to ignore.